srcts/src/components/webcomponents/layoutColumns.ts

```diff
@@ -0,0 +1,543 @@
+    const attr = this.getAttribute("col-widths");
+    if (!attr) {
+      return BslibLayoutColumns.defaultColWidths();
+    }
+
+    let x = JSON.parse(attr);
+    if (Array.isArray(x)) {
+      x = { md: x };
+    }
+
+    const colWidths = new Map();
+    const breaks = ["sm", "md", "lg", "xl", "xxl"];
+
+    // Assign known bootstrap breakpoints in order
+    breaks.forEach((breakName) => {
+      if (x[breakName]) {
+        colWidths.set(breakName, isNA(x[breakName]) ? null : x[breakName]);
+        delete x[breakName];
+      }
+    });
+    // Plus any remaining breakpoints
+    Object.keys(x).forEach((breakName) => {
+      colWidths.set(breakName, isNA(x[breakName]) ? null : x[breakName]);
+    });
+
+    // fill in `md` breakpoint if only large breakpoints are specified
+    const hasLg = ["lg", "xl", "xxl"].some((size) => colWidths.has(size));
+    const hasSm = ["sm", "md"].some((size) => colWidths.has(size));
+
+    if (hasLg && !hasSm) {
+      const smallest = ["lg", "xl", "xxl"].filter((size) =>
+        colWidths.has(size)
+      );
+      colWidths.set("md", colWidths.get(smallest[0]));
+    }
+
+    return colWidths;
+  }
+
+  /**
+   * Given user-specified column widths, resolve the auto-fit breakpoints and
+   * then finalize the column width spec. If any auto-fit breakpoints are
+   * requested, this function needs to know the number of children in the
+   * layout, so it can determine the best-fit column widths.
+   *
+   * @returns The resolved column width specification.
+   */
+  private _resolveColWidthsSpec(): BreakpointColumnSpec {
+    const allResolved = Array.from(this.colWidths.values()).every(
+      (val) => !isNA(val)
+    );
+
+    if (allResolved) {
+      return newBreakpointColumnSpec(this.colWidths as BreakpointMapResolved);
+    }
+
+    const resolved = new Map() as BreakpointMapResolved;
+
+    const allAutoFit = Array.from(this.colWidths.values()).every((val) =>
+      isNA(val)
+    );
+
+    const units = allAutoFit ? null : 12;
+    const nChildren = this.children.length;
+
+    for (const [breakName, colWidth] of this.colWidths) {
+      if (colWidth === null) {
+        const preferWider = ["sm", "md"].includes(breakName);
+        const bestFit = bestFitColumnWidths(nChildren, preferWider, units);
+
+        if (allAutoFit) {
+          this.colUnits = bestFit.units;
+        }
+
+        resolved.set(breakName, bestFit.widths);
+      } else {
+        resolved.set(breakName, colWidth);
+      }
+    }
+
+    return newBreakpointColumnSpec(resolved);
+  }
+
+  /**
+   * Applies the column width specification to the children of the custom element.
+   */
+  private _applyColWidthsSpec(): void {
+    if (!this.colWidthsSpec) {
+      this.colWidthsSpec = this._resolveColWidthsSpec();
+    }
+
+    if (!this.colWidthsSpec) {
+      throw new Error("Column widths must be specified.");
+    }
+
+    const children = this.children;
+    writeGridClasses(this.colWidthsSpec, children, this.colUnits);
```

Comment:
Currently, the children start as "grid items" already, i.e. `div.bslib-grid-item.bslib-gap-spacing` plus `.html-fill-container` if `fillable = TRUE`.

It's better for these items to start wrapped in the grid item containers because it reduces the impact of the custom element failing for some reason -- the markup will already be mostly correct other than the column width classes.

On the other hand, the children could be bare children that are promoted to grid items, if not  already. This would make it easier to use `<bslib-layout-columns>` in contexts where it isn't reasonable to add the grid item container, as well as dynamically adding new contents to the layout (which we don't currently support but could in the future).

Reply:
I don't think we need to do this now. In fact, I adjusted our CSS to not rely on the `.bslib-grid-item` class at all: 425ecfbe9941078f4bf5cad9751db4942cb477e5